tests/integration_tests/reaper.py

```diff
@@ -0,0 +1,198 @@
+destroyed. Instances that refuse to be destroyed due to external library errors
+or flaky infrastructure are tracked, retried and upon test session completion
+are reported to the end user as a test warning.
+"""
+
+import logging
+import queue
+import threading
+import warnings
+from typing import Final, List
+
+from tests.integration_tests.instances import IntegrationInstance
+
+LOG = logging.getLogger()
+
+
+class _Reaper:
+    def __init__(self, timeout: float = 30.0):
+        # self.timeout sets the amount of time to sleep before retrying
+        self.timeout = timeout
+        # self.wake_reaper tells the reaper to wake up.
+        #
+        # A lock is used for synchronization. This means that notify() will
+        # block if
+        # the reaper is currently awake.
+        #
+        # It is set by:
+        # - signal interrupt indicating cleanup
+        # - session completion indicating cleanup
+        # - reaped instance indicating work to be done
+        self.wake_reaper: Final[threading.Condition] = threading.Condition()
+
+        # self.exit_reaper tells the reaper loop to tear down, called once at
+        # end of tests
+        self.exit_reaper: Final[threading.Event] = threading.Event()
+
+        # List of instances which temporarily escaped death
+        # The primary porpose of the reaper is to coax these instance towards
+        # eventual demise and report their insubordination on shutdown.
+        self.undead_ledger: Final[List[IntegrationInstance]] = []
+
+        # Queue of newly reaped instances
+        self.reaped_instances: Final[queue.Queue[IntegrationInstance]] = (
+            queue.Queue()
+        )
+
+        # Thread object, handle used to re-join the thread
+        self.reaper_thread: threading.Thread
+
+        # Count the dead
+        self.counter = 0
+
+    def reap(self, instance: IntegrationInstance):
+        """reap() submits an instance to the reaper thread.
+
+        An instance that is passed to the reaper must not be used again. It may
+        not be dead yet, but it has no place among the living.
+        """
+        LOG.info("Reaper: receiving %s", instance.instance.id)
+
+        self.reaped_instances.put(instance)
+        with self.wake_reaper:
+            self.wake_reaper.notify()
+            LOG.info("Reaper: awakened to reap")
+
+    def start(self):
+        """Spawn the reaper background thread."""
+        LOG.info("Reaper: starting")
+        self.reaper_thread = threading.Thread(
+            target=self._reaper_loop, name="reaper"
+        )
+        self.reaper_thread.start()
+
+    def stop(self):
+        """Stop the reaper background thread and wait for completion."""
+        LOG.info("Reaper: stopping")
+        self.exit_reaper.set()
+        with self.wake_reaper:
+            self.wake_reaper.notify()
+            LOG.info("Reaper: awakened to reap")
+        if self.reaper_thread:
```

Comment:
(suggestion (now deleted) from previous review was on the wrong line)
```suggestion
        if self.reaper_thread and reaper_thread.is_alive():
```